examples/next-openai/app/function-calling/page.tsx

```diff
@@ -0,0 +1,205 @@
+'use client'
+
+import { ChatCompletionFunctions } from 'openai-edge/types/api'
+import { Message } from 'ai/react'
+import { useChat } from 'ai/react'
+import { ChatRequest, FunctionCallHandler, nanoid } from 'ai'
+
+const functions: ChatCompletionFunctions[] = [
+  {
+    name: 'get_current_weather',
+    description: 'Get the current weather',
+    parameters: {
+      type: 'object',
+      properties: {
+        location: {
+          type: 'string',
+          description: 'The city and state, e.g. San Francisco, CA'
+        },
+        format: {
+          type: 'string',
+          enum: ['celsius', 'fahrenheit'],
+          description:
+            'The temperature unit to use. Infer this from the users location.'
+        }
+      },
+      required: ['location', 'format']
+    }
+  },
+  {
+    name: 'get_current_time',
+    description: 'Get the current time',
```

Comment:
Good point. The functions that are passed to OpenAI in a request can also be defined and kept server side (I can update the example to do this), but this PR enables function calling to be handled _client side_ with `onFunctionCall` (which is different than the flows suggested in your [tweet](https://twitter.com/jaredpalmer/status/1673350963191508993), and a use case I'm most excited about). As such, the client has to have some knowledge about the various potential function calls and how to handle them in the `FunctionCallHandler`.

Reply:
In case anyone needs to be sold on client-side handling of function calls, here some examples I'm excited about:

- Evaluating LLM-generated javascript code in the browser with `eval` to create d3.js charts, create/iterate on browser-based games, changing the styling or look and fell of a user's page, etc.. Also, evaluating code that users generate through their prompting in the browser rather than on the server mitigates security risks of evaluating user-generated code server-side where more sensitive resources could be abused.

- Conducting complicated data queries and data analysis of tabular data in the browser with [sqlite/WASM](https://developer.chrome.com/blog/sqlite-wasm-in-the-browser-backed-by-the-origin-private-file-system/) by using LLMs to generate SQL that is executed on the sqlite instance running in the browser.
-

Reply:
I am also using client side function calling, for things like updating the user's name in a chat application, redirecting to other pages, and things like that.  That said, doing server side function calling isn't as easy, as you have to jsut watch the onCompletion and parse any function calls yourself, so we might want another PR to address that.

Reply:
an `onFunctionCall` server-side callback is definitely planned, but we can work on the next steps after this work is merged.

Reply:
Yes, I do think server-side function calling as outlined in Jared's tweet is important to support, and I think it can be fully complementary with client-side function calling

Reply:
@jaredpalmer  I've moved the function definitions in the `examples/next-openai` example to be server-side.

Reply:
I think the server side execution of functions is probably just as important as the client side, but @Zakinator123 brings up some very compelling usecases for public executions.